Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp

```diff
@@ -0,0 +1,646 @@
+    const intType size  = 16;
+    const intType n = size * size * size; // A is n x n
+    
+    const intType nnzUB = 27 * n; // upper bound of nnz from 27 point stencil
+
+    // PCG settings
+    const intType maxIter = 500;
+    const dataType relTol = 1.0e-5;
+    const dataType absTol = 5.0e-4;
+
+    // Catch asynchronous exceptions
+    auto exception_handler = [](sycl::exception_list exceptions) {
+        for (std::exception_ptr const &e : exceptions) {
+            try {
+                std::rethrow_exception(e);
+            }
+            catch (sycl::exception const &e) {
+                std::cout << "Caught asynchronous SYCL "
+                             "exception during sparse CG:\n"
+                          << e.what() << std::endl;
+            }
+        }
+    };
+
+    // create queue around device
+    sycl::queue q(dev, exception_handler);
+
+
+    // Input matrix in CSR format
+    intType *ia_h = sycl::malloc_host<intType>(n+1, q);
+    intType *ja_h = sycl::malloc_host<intType>(nnzUB, q);
+    dataType *a_h = sycl::malloc_host<dataType>(nnzUB, q);
+    dataType *x_h = sycl::malloc_host<dataType>(n, q);
+    dataType *b_h = sycl::malloc_host<dataType>(n, q);
+
+    if (!ia_h || !ja_h || !a_h || !x_h || !b_h ) {
+        throw std::runtime_error("Failed to allocate host side USM memory");
+    }
+
+    // 
+    // Generate a 27 point stencil for 3D laplacian using size elements in each dimension
+    //
+    generate_sparse_matrix<dataType, intType>(size, ia_h, ja_h, a_h);
+
+    // Init right hand side and vector x
+    for (int i = 0; i < n; i++) {
+        b_h[i] = set_fp_value(dataType(1.0), dataType(0.0)); // rhs b = 1
+        x_h[i] = set_fp_value(dataType(0.0), dataType(0.0)); // initial guess x0 = 0
+    }
+
+    //
+    // Execute Preconditioned Conjugate Gradient Algorithm
+    //
+    // solve  A x = b  starting with initial guess x = x0
+    //
+
+    std::cout << "\n\t\tsparse PCG parameters:\n";
+
+    std::cout << "\t\t\tA size: (" << n << ", " << n << ")" << std::endl;
+    std::cout << "\t\t\tPreconditioner = Symmetric Gauss-Seidel" << std::endl;
+    std::cout << "\t\t\tmax iterations = " << maxIter << std::endl;
+    std::cout << "\t\t\trelative tolerance limit = " << relTol << std::endl;
+    std::cout << "\t\t\tabsolute tolerance limit = " << absTol << std::endl;
+
+
+    const intType nnz = ia_h[n]; // assumes zero indexing
+
+    // create arrays for help
+    intType *ia_d    = sycl::malloc_device<intType>(n+1, q);
+    intType *ja_d    = sycl::malloc_device<intType>(nnz, q);
+    dataType *a_d    = sycl::malloc_device<dataType>(nnz, q);
+    dataType *x_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *b_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *r_d    = sycl::malloc_device<dataType>(n, q); // r = b - A*x
+    dataType *z_d    = sycl::malloc_device<dataType>(n, q); // z = M^{-1} * r
+    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = ?? z
```

Comment:
```suggestion
    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = z + beta * p
```
Or "search direction"?